refactor: unit tree component

Pucek9 opened the pull request:
Closes #8199

tobi-or-not-tobi committed:
Add new styles

tobi-or-not-tobi committed:
small style fixes

tobi-or-not-tobi committed:
small fixes

tobi-or-not-tobi committed:
add budget to public api

dunqan committed:
fix imports for prod build

dunqan committed:
fix imports for specs

tobi-or-not-tobi committed:
fix tests

tobi-or-not-tobi committed:
small improvements

Commit:
fix: disable special characters for id fields (#8822)

Closes: #8606

tobi-or-not-tobi committed:
message improvements

Commit:
chore: provide missing unit test files for list services (#8776)

Closes: #8417

tobi-or-not-tobi committed:
fix tests

tobi-or-not-tobi committed:
fix production build

tobi-or-not-tobi committed:
Add permissions

tobi-or-not-tobi committed:
User group UX refactor

tobi-or-not-tobi committed:
Align budget with new sub list structure

tobi-or-not-tobi committed:
align cost center with new sub list structure

tobi-or-not-tobi committed:
restore order approval

szczepaniak-michal committed:
Feature/gh 8194 (#8435)

Closes #8194

tobi-or-not-tobi committed:
Convert user area

tobi-or-not-tobi committed:
missing locales

tobi-or-not-tobi committed:
stacked messages

szczepaniak-michal committed:
GH-8194: Fix route param codes

tobi-or-not-tobi committed:
message stack and termination

tobi-or-not-tobi committed:
drop hovering on non table anchor links

tobi-or-not-tobi committed:
anchor links instead of buttons

tobi-or-not-tobi committed:
move password

tobi-or-not-tobi committed:
column based fieldset

tobi-or-not-tobi committed:
move budget edit back to right view position...

tobi-or-not-tobi committed:
drop logging

tobi-or-not-tobi committed:
clean up form html

tobi-or-not-tobi committed:
renaming

tobi-or-not-tobi committed:
locales

Pucek9 committed:
init

Pucek9 committed:
up

tobi-or-not-tobi committed:
split view gutter is back

tobi-or-not-tobi committed:
Add units

tobi-or-not-tobi committed:
Unit Details

tobi-or-not-tobi committed:
unit form component

tobi-or-not-tobi committed:
parent unit in form

Pucek9 committed:
Merge branch 'feature/GH-8704-nextgen' of https://github.com/SAP/cloud-commerce-spartacus-storefront into feature/GH-8704-nextgen-units

# Conflicts:
#	feature-libs/my-account/organization/assets/translations/en/units.i18n.ts
#	feature-libs/my-account/organization/components/shared/organization-detail/toggle-status-action/toggle-status.component.ts
#	feature-libs/my-account/organization/components/shared/organization.model.ts
#	feature-libs/my-account/organization/components/unit/addresses/create/unit-address-create.component.html
#	feature-libs/my-account/organization/components/unit/addresses/create/unit-address-create.component.spec.ts
#	feature-libs/my-account/organization/components/unit/addresses/create/unit-address-create.component.ts
#	feature-libs/my-account/organization/components/unit/addresses/details/current-unit-address.service.spec.ts
#	feature-libs/my-account/organization/components/unit/addresses/details/current-unit-address.service.ts
#	feature-libs/my-account/organization/components/unit/addresses/details/unit-address-details.component.html
#	feature-libs/my-account/organization/components/unit/addresses/details/unit-address-details.component.spec.ts
#	feature-libs/my-account/organization/components/unit/addresses/details/unit-address-details.component.ts
#	feature-libs/my-account/organization/components/unit/addresses/edit/unit-address-edit.component.html
#	feature-libs/my-account/organization/components/unit/addresses/edit/unit-address-edit.component.spec.ts
#	feature-libs/my-account/organization/components/unit/addresses/edit/unit-address-edit.component.ts
#	feature-libs/my-account/organization/components/unit/addresses/form/unit-address-form.module.ts
#	feature-libs/my-account/organization/components/unit/addresses/list/unit-address-list.component.html
#	feature-libs/my-account/organization/components/unit/addresses/list/unit-address-list.service.ts
#	feature-libs/my-account/organization/components/unit/approvers/assign/unit-assign-approvers.component.html
#	feature-libs/my-account/organization/components/unit/approvers/assign/unit-assign-approvers.service.ts
#	feature-libs/my-account/organization/components/unit/approvers/list/unit-approver-list.component.html
#	feature-libs/my-account/organization/components/unit/approvers/list/unit-approver-list.service.ts
#	feature-libs/my-account/organization/components/unit/children/unit-children.component.html
#	feature-libs/my-account/organization/components/unit/children/unit-children.service.ts
#	feature-libs/my-account/organization/components/unit/cost-centers/unit-cost-centers.component.html
#	feature-libs/my-account/organization/components/unit/cost-centers/unit-cost-centers.service.ts
#	feature-libs/my-account/organization/components/unit/current-unit.service.spec.ts
#	feature-libs/my-account/organization/components/unit/current-unit.service.ts
#	feature-libs/my-account/organization/components/unit/details/unit-details.component.html
#	feature-libs/my-account/organization/components/unit/details/unit-details.component.ts
#	feature-libs/my-account/organization/components/unit/details/unit-details.module.ts
#	feature-libs/my-account/organization/components/unit/form/unit-form.component.html
#	feature-libs/my-account/organization/components/unit/form/unit-form.component.spec.ts
#	feature-libs/my-account/organization/components/unit/form/unit-form.component.ts
#	feature-libs/my-account/organization/components/unit/form/unit-form.module.ts
#	feature-libs/my-account/organization/components/unit/form/unit-form.service.ts
#	feature-libs/my-account/organization/components/unit/index.ts
#	feature-libs/my-account/organization/components/unit/list/unit-list.component.html
#	feature-libs/my-account/organization/components/unit/list/unit-list.component.spec.ts
#	feature-libs/my-account/organization/components/unit/list/unit-list.component.ts
#	feature-libs/my-account/organization/components/unit/services/unit-item.service.ts
#	feature-libs/my-account/organization/components/unit/units-components.module.ts
#	feature-libs/my-account/organization/components/unit/units.config.ts
#	feature-libs/my-account/organization/components/unit/users/assign-roles/unit-user-assign-roles.component.html
#	feature-libs/my-account/organization/components/unit/users/assign-roles/unit-user-assign-roles.service.ts
#	feature-libs/my-account/organization/components/unit/users/list/unit-user-list.component.html
#	feature-libs/my-account/organization/components/unit/users/list/unit-users.service.ts

Pucek9 committed:
revert preffixes removing

Pucek9 committed:
fix

Pucek9 committed:
remove unnecessary files

tobi-or-not-tobi committed:
add user

tobi-or-not-tobi committed:
Add missing active state to unit details

tobi-or-not-tobi committed:
Merge branch 'feature/GH-8704-nextgen' into feature/GH-8704-nextgen-units

tobi-or-not-tobi committed:
Unit approver added

Pucek9 committed:
fix translation

tobi-or-not-tobi committed:
user roles

tobi-or-not-tobi committed:
unit cost centers

Pucek9 committed:
up

tobi-or-not-tobi committed:
fix prod build

tobi-or-not-tobi committed:
localized role change messages

Pucek9 committed:
up

Pucek9 committed:
up

Pucek9 committed:
up

Pucek9 committed:
up

tobi-or-not-tobi committed:
unit addresses

Pucek9 committed:
Merge branch 'feature/GH-8704-nextgen' of https://github.com/SAP/cloud-commerce-spartacus-storefront into feature/GH-8704-nextgen-units

# Conflicts:
#	feature-libs/my-account/organization/components/unit/services/unit-list.service.ts
#	feature-libs/my-account/organization/components/unit/units.config.ts

tobi-or-not-tobi committed:
fix split view sizing

Pucek9 committed:
Merge branch 'feature/GH-8704-nextgen' of https://github.com/SAP/cloud-commerce-spartacus-storefront into feature/GH-8704-nextgen-units

tobi-or-not-tobi committed:
update split view state based on screen size

tobi-or-not-tobi committed:
drop mock

tobi-or-not-tobi committed:
outdated

Pucek9 committed:
Merge branch 'feature/GH-8704-nextgen' of https://github.com/SAP/cloud-commerce-spartacus-storefront into feature/GH-8704-nextgen-units

tobi-or-not-tobi committed:
save user roles

Matejk00 committed:
fix: data loading on assign pages (#8644)

Closes #8332

tobi-or-not-tobi committed:
user form adjustments

Pucek9 committed:
Merge branch 'feature/GH-8704-nextgen' of https://github.com/SAP/cloud-commerce-spartacus-storefront into feature/GH-8704-nextgen-units

tobi-or-not-tobi committed:
prod build

tobi-or-not-tobi committed:
fix storefront tests

tobi-or-not-tobi committed:
align tests

tobi-or-not-tobi committed:
add tests for sub list

tobi-or-not-tobi committed:
improve cell components + adding no link cell feature

tobi-or-not-tobi committed:
drop unused service

tobi-or-not-tobi committed:
add unit to public api

tobi-or-not-tobi committed:
Merge branch 'feature/my-company' into feature/GH-8704-nextgen

tobi-or-not-tobi committed:
missing service

Pucek9 committed:
up

tobi-or-not-tobi committed:
fix tests

Pucek9 committed:
temporary workaround

Pucek9 committed:
up with safe operator

tobi-or-not-tobi committed:
move edit to separate view

tobi-or-not-tobi committed:
combine currency and amount

Pucek9 committed:
Merge branch 'feature/GH-8704-nextgen' of https://github.com/SAP/cloud-commerce-spartacus-storefront into feature/GH-8704-nextgen-units

# Conflicts:
#	feature-libs/my-account/organization/assets/translations/en/units.i18n.ts
#	feature-libs/my-account/organization/components/public_api.ts

tobi-or-not-tobi committed:
add missing AdminGuard

tobi-or-not-tobi committed:
cleanup user group locales

Pucek9 committed:
up

Pucek9 committed:
Merge branch 'feature/GH-8704-nextgen' of https://github.com/SAP/cloud-commerce-spartacus-storefront into feature/GH-8704-nextgen-units

tobi-or-not-tobi committed:
missing year permissin

Pucek9 committed:
Merge branch 'feature/GH-8704-nextgen' of https://github.com/SAP/cloud-commerce-spartacus-storefront into feature/GH-8704-nextgen-units

Pucek9 committed:
up

Pucek9 committed:
up

tobi-or-not-tobi committed:
fixed overflowng of nested form selects with large text

tobi-or-not-tobi committed:
Update feature-libs/my-account/organization/assets/translations/translations.ts

Co-authored-by: Michał Gruca <pucek9@gmail.com>

tobi-or-not-tobi committed:
Merge branch 'feature/GH-8704-nextgen' of https://github.com/SAP/spartacus into feature/GH-8704-nextgen

tobi-or-not-tobi committed:
fix form layout issue

tobi-or-not-tobi committed:
revert core service

Pucek9 committed:
Merge branch 'feature/GH-8704-nextgen' of https://github.com/SAP/cloud-commerce-spartacus-storefront into feature/GH-8704-nextgen-units

Pucek9 committed:
fix routing for work with current mechanism

Pucek9 committed:
fix for hiding grand children

Pucek9 committed:
update

Pucek9 committed:
finally worked

Pucek9 committed:
expand branch with active route

Pucek9 committed:
up

Pucek9 committed:
add abstract action class

tobi-or-not-tobi committed:
Allow for text overflow on list items

tobi-or-not-tobi commented on feature-libs/my-account/organization/components/shared/organization-list/organization-list.component.html:
```diff
@@ -25,6 +25,8 @@ <h3>
             </ng-option>
           </ng-select>
 
+          <ng-content></ng-content>
```
I'd rather use the explicit projection that we used in the other shared components, i.e. in `organization-component.html`:

```html
<ng-content select="[actions]"></ng-content>
```

tobi-or-not-tobi committed:
fix getting table cell options

tobi-or-not-tobi committed:
introduce accessibility focus for main list

tobi-or-not-tobi committed:
introduce message notification with default notification removal

tobi-or-not-tobi committed:
keyboarding on notifications and prompts

tobi-or-not-tobi committed:
soft-close on cancel

tobi-or-not-tobi committed:
leverage new notify message

Commit:
fix: change to mergeMap to avoid cancelling previous entity assignments (#8840)

Closes: #8380

tobi-or-not-tobi committed:
avoid link on linked list

tobi-or-not-tobi committed:
avoid editing permissions assigned to inactive unit

tobi-or-not-tobi committed:
fix locales

Pucek9 committed:
add select=[actions]

Pucek9 committed:
update config for mobile devices

tobi-or-not-tobi committed:
style optimisation

Pucek9 committed:
fix uid hover cursor

Pucek9 committed:
Merge branch 'feature/GH-8704-nextgen' of https://github.com/SAP/cloud-commerce-spartacus-storefront into feature/GH-8704-nextgen-units

Pucek9 committed:
hide button for root unit

Pucek9 committed:
fix issue with collapsed root

Pucek9 committed:
mobile styling

tobi-or-not-tobi committed:
unit roles assignment

tobi-or-not-tobi committed:
drop log

Pucek9 committed:
Merge branch 'feature/GH-8704-nextgen' of https://github.com/SAP/cloud-commerce-spartacus-storefront into feature/GH-8704-nextgen-units

# Conflicts:
#	feature-libs/my-account/organization/components/unit/units.config.ts

Pucek9 committed:
change icons

tobi-or-not-tobi committed:
finalize company landing page

Pucek9 committed:
Merge branch 'feature/GH-8704-nextgen' of https://github.com/SAP/cloud-commerce-spartacus-storefront into feature/GH-8704-nextgen-units

tobi-or-not-tobi committed:
missed margin

dunqan committed:
Sync my-company with develop (#8934)

Resolving conflicts after merging b2b-checkout to develop

tobi-or-not-tobi committed:
dropping obsolete selectors

tobi-or-not-tobi committed:
Merge branch 'feature/my-company' into feature/GH-8704-nextgen

Pucek9 committed:
Merge branch 'feature/GH-8704-nextgen' of https://github.com/SAP/cloud-commerce-spartacus-storefront into feature/GH-8704-nextgen-units

Pucek9 committed:
change place of providers

Pucek9 committed:
fix

Pucek9 committed:
Merge branch 'feature/GH-8704-nextgen' of https://github.com/SAP/cloud-commerce-spartacus-storefront into feature/GH-8704-nextgen-units

Pucek9 committed:
prettier fix

dunqan committed:
build fixes

tobi-or-not-tobi committed:
fix test

tobi-or-not-tobi committed:
fix tests

tobi-or-not-tobi committed:
Merge branch 'feature/GH-8704-nextgen' into feature/GH-8704-nextgen-units

tobi-or-not-tobi commented on feature-libs/my-account/organization/components/shared/organization-table/toggle-link/toggle-link-cell.component.ts:
```diff
@@ -0,0 +1,62 @@
+
+@Component({
+  template: `
+    <ng-container *ngIf="count > 0 && nestLevel > 0; else noExpandable">
```
Can we move the template in a separate file?

tobi-or-not-tobi commented on feature-libs/my-account/organization/components/shared/organization-table/toggle-link/toggle-link-cell.component.ts:
```diff
@@ -0,0 +1,62 @@
+  `,
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class ToggleLinkCellComponent extends OrganizationCellComponent {
```
Can we move the component to the unit folder? It seems only used/relevant to unit tree.

tobi-or-not-tobi commented on feature-libs/my-account/organization/components/shared/organization-table/toggle-link/toggle-link-cell.component.ts:
```diff
@@ -0,0 +1,62 @@
+})
+export class ToggleLinkCellComponent extends OrganizationCellComponent {
+  @HostBinding('style.--cx-nest-level')
+  nestLevel = this.model.depthLevel;
```
hostbinding should work fine on a getter as well. Either way we should use a single style (property vs getter method). I have used getters as I'd expect changes to be fetched if the outlet context is changing, while the component doesn't update. I might be wrong.

tobi-or-not-tobi replied:
we'd need to align with Miguel regarding accessibility angle of the expand/ collapse

tobi-or-not-tobi commented on feature-libs/my-account/organization/components/shared/organization-table/toggle-link/toggle-link-cell.component.ts:
```diff
@@ -0,0 +1,62 @@
+@Component({
+  template: `
+    <ng-container *ngIf="count > 0 && nestLevel > 0; else noExpandable">
+      <button class="button action" (click)="toggleItem($event)">
```
would be nice to simplify this: 

```html
<button
      *ngIf="count > 0 && nestLevel > 0"
      class="button tree-item-toggle"
      type="button"
      (click)="toggleItem($event)"
    >
      <cx-icon [type]="expanded ? 'CARET_DOWN' : 'CARET_RIGHT'"></cx-icon>
</button>
```

The button type is recommended as technically the button could otherwise submit  a form.

tobi-or-not-tobi replied:
I'd use a bit more semantic  class for the button, would also be good to add expand vs collapse as a class name.

tobi-or-not-tobi replied:
In the styling we should make the button full height/width of the cell, so that it's easily clickable.

tobi-or-not-tobi commented on feature-libs/my-account/organization/styles/__ux/_list.scss:
```diff
@@ -287,6 +287,13 @@ cx-organization-form {
+    td.name {
+      button.button.action {
+        align-self: center;
+        margin-left: calc(var(--cx-nest-level) * 1.5rem);
```
Please use logical properties to ensure we're bidirectional.

tobi-or-not-tobi replied:
nice one with the css var

tobi-or-not-tobi replied:
`margin-inline-start: calc(var(--cx-nest-level) * 1.5rem);`

tobi-or-not-tobi commented on feature-libs/my-account/organization/styles/__ux/_list.scss:
```diff
@@ -379,3 +386,26 @@ cx-organization-form {
+        align-self: center;
+      }
+
+      button.button.action {
```
lengthy selector. I'd like a more semantic selector, i.e. `toggle-tree` and `expand` or `collapse`.

tobi-or-not-tobi commented on feature-libs/my-account/organization/styles/__ux/_list.scss:
```diff
@@ -379,3 +386,26 @@ cx-organization-form {
 
 // }
+cx-unit-list {
+  @include media-breakpoint-down(md) {
```
we need to revise the css as the right table layout seems the VERTICAL layout. The VERTICAL_STACKED layout that we use by default for the list component doesn't make to much sense here.

tobi-or-not-tobi commented on feature-libs/my-account/organization/components/unit/units.config.ts:
```diff
@@ -197,6 +200,25 @@ export function unitsTableConfigFactory(): TableConfig {
   table: {
+    [OrganizationTableType.UNIT]: {
+      cells: ['name'],
+      options: {
```
Use the vertical layout for all screens: 

`layout: TableLayout.VERTICAL`

This will avoid the TH in each row.

tobi-or-not-tobi commented on feature-libs/my-account/organization/components/unit/list/unit-list.component.ts:
```diff
@@ -1,37 +1,19 @@
+import { UnitListService } from '../services/unit-list.service';
 
 @Component({
   selector: 'cx-unit-list',
```
this selector can be dropped

tobi-or-not-tobi commented on feature-libs/my-account/organization/components/unit/services/unit-list.service.ts:
```diff
@@ -16,61 +118,88 @@ import { OrganizationTableType } from '../../shared/organization.model';
+    protected unitItemService: UnitItemService
   ) {
     super(tableService);
+    this.unitItemService.key$
```
we must move the logic to a method, so it's extensible.

tobi-or-not-tobi commented on feature-libs/my-account/organization/components/unit/services/unit-list.service.ts:
```diff
@@ -16,61 +118,88 @@ import { OrganizationTableType } from '../../shared/organization.model';
     super(tableService);
+    this.unitItemService.key$
+      .pipe(
+        take(1),
```
You'd need `first((key) => Boolean(key))` instead of `take(1)` and `filter`.

tobi-or-not-tobi commented on feature-libs/my-account/organization/components/unit/services/unit-list.service.ts:
```diff
@@ -16,61 +118,88 @@ import { OrganizationTableType } from '../../shared/organization.model';
+        take(1),
+        filter((key) => Boolean(key))
+      )
+      .subscribe((id) =>
```
You've introduced a memory leak.

tobi-or-not-tobi replied:
you probably better use the `key$` as an inner observable of the `treeAction$` so that you don't need to subscribe at all.

tobi-or-not-tobi commented on feature-libs/my-account/organization/components/unit/services/unit-list.service.ts:
```diff
@@ -16,61 +118,88 @@ import { OrganizationTableType } from '../../shared/organization.model';
+      );
   }
 
+  isExpandedNodeMap = {};
```
I think you need to go over all accessors of properties and methods

Pucek9 committed:
Merge branch 'feature/GH-8704-nextgen' of https://github.com/SAP/cloud-commerce-spartacus-storefront into feature/GH-8704-nextgen-units

Pucek9 committed:
Merge branch 'feature/GH-8704-nextgen-units' of https://github.com/SAP/cloud-commerce-spartacus-storefront into feature/GH-8704-nextgen-units

tobi-or-not-tobi committed:
Restructure CSS for my company

tobi-or-not-tobi committed:
forgotten class

Pucek9 committed:
Merge branch 'feature/GH-8704-nextgen' of https://github.com/SAP/cloud-commerce-spartacus-storefront into feature/GH-8704-nextgen-units

# Conflicts:
#	feature-libs/my-account/organization/styles/__ux/_list.scss

Pucek9 committed:
Merge branch 'feature/GH-8704-nextgen' of https://github.com/SAP/cloud-commerce-spartacus-storefront into feature/GH-8704-nextgen-units

Pucek9 committed:
update

Pucek9 committed:
update

Pucek9 committed:
styling update

Pucek9 committed:
styling update

Pucek9 committed:
add accessor for expandedNodeMap

Pucek9 committed:
move stream to method

tobi-or-not-tobi committed:
simplified tree logic

tobi-or-not-tobi committed:
improved message component

Pucek9 committed:
fix: General refresh issue after editing or creating new things #8518 (#8816)

Closes: #8518

Platonn committed:
Sync my-account with develop (#8963)

Commits from develop:
ed97b9a26 chore: avoid duplicated config in b2b.feature and b2c.feature (#8953)
69cab03a6 GH-8915 Fixing e2e mobile search test
73c2859ab Release/3.0.0 next.1 (#8950)
95cf8e405 Updated "module" compiler options to es2020 (#8938)
95e168603 test: basic e2e for b2b checkout (#8912)

Platonn committed:
Merge branch 'develop' of github.com:SAP/cloud-commerce-spartacus-storefront into feature/my-company

Pucek9 committed:
add expand all

Pucek9 committed:
collapse all

Pucek9 committed:
add pagination total results

Pucek9 committed:
branch expanded by url key

tobi-or-not-tobi committed:
improved message component + tests

tobi-or-not-tobi committed:
drop async in tests setup

tobi-or-not-tobi committed:
Merge branch 'feature/my-company' into feature/GH-8704-nextgen

tobi-or-not-tobi committed:
Merge branch 'feature/GH-8704-nextgen' into feature/GH-8704-nextgen-units

tobi-or-not-tobi commented on feature-libs/my-account/organization/components/unit/services/unit-list.service.ts:
```diff
@@ -17,60 +25,118 @@ import { OrganizationTableType } from '../../shared/organization.model';
-      .pipe(map((raw) => this.convertUnits(raw)));
+      .pipe(
+        withLatestFrom(this.unitItemService.key$),
+        first(([_tree, id]) => Boolean(id))
```
Indeed no memory leak here 👍  
I'd still avoid the explicit subscribe here, and use a switchAll inside the load method, to go from the initial key to loading getTree. But nothing fundamental wrong here.

tobi-or-not-tobi commented on feature-libs/my-account/organization/components/unit/services/unit-list.service.ts:
```diff
@@ -17,60 +25,118 @@ import { OrganizationTableType } from '../../shared/organization.model';
+    this.levelState.set(treeNode.uid, depthLevel);
+
+    values.push(treeNode);
+    pagination.totalResults++;
```
I was under the assumption that we don't have pagination in the tree component, or am I wrong?

Pucek9 replied:
Yes, we don't have pagination here, but 'pagination' object is used in list header, I decided for wrapping it to be compatible with current components (simulate / workaround backend output).

tobi-or-not-tobi commented on feature-libs/my-account/organization/components/unit/services/unit-list.service.ts:
```diff
@@ -17,60 +25,118 @@ import { OrganizationTableType } from '../../shared/organization.model';
+  protected initialExpanded = 1;
+
+  protected expandState = new Map<string, boolean>();
+  protected levelState = new Map<string, number>();
```
I would not use the map to store all levels. I'd rather maintain a single value for the expanded/collapsed state, and take that into account when building the tree. I think it's easier and clear, as you could potentially rollback the levelState, and with that return to the previous expanded state.

Pucek9 replied:
I hope it would be what you like: [`f2b5f8d` (#8894)](https://github.com/SAP/spartacus/pull/8894/commits/f2b5f8dc63337ee23921e5fe251d2e51c4d7316f)

tobi-or-not-tobi reviewed (changes requested):
Some suggestions for the `UnitListService`, nothing that stops us from merging actually. 
Are you writing tests as well?

Pucek9 committed:
replace 2xMap to 1 which collects whole objects

Pucek9 committed:
Merge remote-tracking branch 'origin/feature/GH-8704-nextgen-units' into feature/GH-8704-nextgen-units

Pucek9 committed:
styling & cleanup

tobi-or-not-tobi committed:
Unit test for toggle status

tobi-or-not-tobi committed:
fix locales

tobi-or-not-tobi committed:
Merge branch 'feature/GH-8704-nextgen' into feature/GH-8704-nextgen-units

tobi-or-not-tobi committed:
fix link button text color

tobi-or-not-tobi committed:
align links on list pages

tobi-or-not-tobi committed:
Merge branch 'feature/GH-8704-nextgen' into feature/GH-8704-nextgen-units

tobi-or-not-tobi committed:
Add edit class selector

Pucek9 committed:
up

tobi-or-not-tobi committed:
Feature/gh 8704 nextgen (#8924)

New UX for my company

Pucek9 committed:
Merge branch 'feature/my-company' of https://github.com/SAP/cloud-commerce-spartacus-storefront into feature/GH-8704-nextgen-units

# Conflicts:
#	feature-libs/my-account/organization/assets/translations/en/units.i18n.ts
#	feature-libs/my-account/organization/components/budget/details/budget-details.component.html
#	feature-libs/my-account/organization/components/cost-center/details/cost-center-details.component.html
#	feature-libs/my-account/organization/components/permission/details/permission-details.component.html
#	feature-libs/my-account/organization/components/shared/organization-list/organization-list.component.html
#	feature-libs/my-account/organization/components/unit/details/unit-details.component.html
#	feature-libs/my-account/organization/components/unit/index.ts
#	feature-libs/my-account/organization/components/unit/links/addresses/details/unit-address-details.component.html
#	feature-libs/my-account/organization/components/unit/list/unit-list.component.spec.ts
#	feature-libs/my-account/organization/components/unit/list/unit-list.component.ts
#	feature-libs/my-account/organization/components/unit/services/unit-item.service.ts
#	feature-libs/my-account/organization/components/unit/services/unit-list.service.ts
#	feature-libs/my-account/organization/components/unit/units.config.ts
#	feature-libs/my-account/organization/components/user-group/details/user-group-details.component.html
#	feature-libs/my-account/organization/components/user/details/user-details.component.html
#	feature-libs/my-account/organization/styles/_index.scss
#	feature-libs/my-account/organization/styles/_list.scss
#	feature-libs/my-account/organization/styles/_tree.scss

Pucek9 committed:
fix unit test of organization-list.component.spec.ts

Platonn committed:
fix: drop obsolete organization translations from the @spartacus/assets lib

Recently (https://github.com/SAP/spartacus/commit/ed97b9a26727ca5c72b27d5bd14bf090db46d41f) we changed the way of merging translations:
- from: spread on the `en` level: `{...myOrgTranslations.en, ...coreTranslations.en}`
- to: the deep merge performed thanks to `ConfigModule.withConfig()` moved to app.module for core translations (+ `.withConfig()` staying in `b2b.feature.ts` for myOrg translations).

This led to a side-effect bug. We had an outdated copy of myOrg translations yet in the `@spartacus/assets` library, which caused conflicts during deep-merge and caused runtime missing translations. 

Now the obsolete assets are removed.

fixes #8985

tobi-or-not-tobi committed:
align budget locales

Pucek9 committed:
Merge branch 'feature/my-company' of https://github.com/SAP/cloud-commerce-spartacus-storefront into feature/GH-8704-nextgen-units

Pucek9 committed:
fix unit test of organization-list.component.spec.ts

Pucek9 committed:
unit-list.component.spec.ts

Pucek9 committed:
initial unit-list.service.spec.ts

Pucek9 committed:
unit-list.service.spec.ts

marlass committed:
chore: Include scss files on lib build (#8992)

Closes #8982

tobi-or-not-tobi commented on feature-libs/my-account/organization/components/unit/services/unit-list.service.ts:
```diff
@@ -1,76 +1,136 @@
   providedIn: 'root',
 })
-export class UnitListService extends OrganizationListService<B2BUnit> {
+export class UnitListService extends OrganizationListService<B2bUnitTreeNode> {
```
While reviewing I've spotted a few issues/limitations with the unit tree:
- (issue) The active tree item is not expanded if you navigate browser history
- (issue) The active item is also expanded, but it shouldn't imho. we should only open the tree until the active item instead. 
- (limitation) We could think of collapsing/expanding _all_ items for a specific part of the tree. This would mean that it's another state 
- (refactor) I'd rather not traverse the full tree map all time when toggling items, it's not really needed. 
- (refactor) The list service becomes a bit busy, i'd move tree related stuff to tree service.

tobi-or-not-tobi replied:
I'd also be in favour of changing the weird behaviour of:
- not having a toggle icon on the root item
- not collapsing all tree nodes, when using collapse all; it's a confusing experience.

tobi-or-not-tobi replied:
And:
- (ux) collapsing all tree nodes, including root children
- (configurability) make the (initial) `minimalExpanded` children extensible.

tobi-or-not-tobi replied:
I've done an optimisation at #9007

Pucek9 replied:
> (issue) The active tree item is not expanded if you navigate browser history

thats a good point, didnt verify that before, sorry

> (issue) The active item is also expanded, but it shouldn't imho. we should only open the tree until the active item instead.

yeah, that was unexpected. Issue probably created while last refactor

> not having a toggle icon on the root item
> not collapsing all tree nodes, when using collapse all; it's a confusing experience.
> (ux) collapsing all tree nodes, including root children

Yeah, I don't like it too but it was done in accordance with business requirements

> (configurability) make the (initial) minimalExpanded children extensible.

I think it was done by `initialExpanded = 1;` property

tobi-or-not-tobi committed:
Introduce tree service, fix a few issues (#9007)

tobi-or-not-tobi committed:
fix button color

tobi-or-not-tobi committed:
make clutter line 100% height when tree is collapsed

Pucek9 committed:
update unit tests

Pucek9 committed:
Merge branch 'feature/GH-8704-nextgen-units' of https://github.com/SAP/cloud-commerce-spartacus-storefront into feature/GH-8704-nextgen-units

Pucek9 committed:
init tests

tobi-or-not-tobi committed:
fix collapsing initial opened nodes

tobi-or-not-tobi committed:
fix style issue on cost-centter->budgets

tobi-or-not-tobi committed:
misconfig

tobi-or-not-tobi committed:
drop obsolete styles

tobi-or-not-tobi committed:
MOve organization root route to organization library

tobi-or-not-tobi committed:
left over

tobi-or-not-tobi committed:
revert incorrect locale

Commit:
fix: review the public api of the my-account/organization entry point (#9032)

Closes: #8571

Pucek9 committed:
Merge branch 'develop' of https://github.com/SAP/cloud-commerce-spartacus-storefront into feature/my-company

# Conflicts:
#	projects/core/src/occ/services/occ-endpoints.service.ts
#	projects/core/src/util/index.ts

marlass committed:
fix: Ghost design on lists (#9056)

Co-authored-by: Marcin Lasak <lasakmarcin90@gmail.com>
Co-authored-by: tobi-or-not-tobi <tobi-or-not-tobi@users.noreply.github.com>
Co-authored-by: Nikola Zarić <nikola.zaric.ns@gmail.com>

marlass committed:
Merge branch 'develop' into feature/my-company

marlass committed:
chore: Remove unused abstract form components

Pucek9 commented on feature-libs/my-account/organization/components/shared/organization-list/organization-list.component.html:
```diff
@@ -49,7 +51,7 @@ <h3>
       <div class="footer">
         <cx-pagination
-          *ngIf="dataTable.pagination"
+          *ngIf="dataTable.pagination?.currentPage"
```
tactical dot

Pucek9 replied:
```suggestion
          *ngIf="dataTable.pagination?.currentPage !== undefined"
```

marlass committed:
fix: Remove default config from public API

Matejk00 committed:
Add unit tests for services

Pucek9 committed:
Merge branch 'develop' of https://github.com/SAP/cloud-commerce-spartacus-storefront into feature/my-company

Matejk00 committed:
Up to develop

Matejk00 committed:
Add suggested pagination statement

Matejk00 committed:
Add missing import for B2BUnitNode

Matejk00 committed:
Update import

Matejk00 committed:
Merge branch 'develop' into feature/GH-8704-nextgen-units

Matejk00 committed:
Up to develop

Matejk00 committed:
Merge branch 'feature/GH-8704-nextgen-units' of https://github.com/SAP/spartacus into feature/GH-8704-nextgen-units

Matejk00 committed:
Move B2BUnitNode to org lib

Matejk00 committed:
Revert pagination statement change

Matejk00 committed:
Show pagination if current page equals to 0

Matejk00 committed:
Merge branch 'develop' into feature/GH-8704-nextgen-units

Pucek9 committed:
removed unused method

Pucek9 committed:
Merge branch 'develop' of https://github.com/SAP/cloud-commerce-spartacus-storefront into feature/GH-8704-nextgen-units

# Conflicts:
#	feature-libs/my-account/organization/styles/_index.scss
#	feature-libs/my-account/organization/styles/_list.scss

Matejk00 committed:
Merge branch 'develop' into feature/GH-8704-nextgen-units

Matejk00 committed:
Merge branch 'develop' into feature/GH-8704-nextgen-units

Matejk00 committed:
Merge branch 'develop' into feature/GH-8704-nextgen-units

Pucek9 committed:
Merge branch 'develop' of https://github.com/SAP/cloud-commerce-spartacus-storefront into feature/GH-8704-nextgen-units

Matejk00 committed:
Merge branch 'develop' into feature/GH-8704-nextgen-units

Pucek9 committed:
Merge branch 'develop' of https://github.com/SAP/cloud-commerce-spartacus-storefront into feature/GH-8704-nextgen-units

# Conflicts:
#	feature-libs/my-account/organization/components/shared/organization-list/organization-list.component.html
#	feature-libs/my-account/organization/components/shared/organization-list/organization-list.component.spec.ts
#	feature-libs/my-account/organization/components/unit/links/approvers/assigned/unit-assigned-approver-list.component.html
#	feature-libs/my-account/organization/components/unit/links/approvers/unit-approver-list.component.html

Pucek9 committed:
Merge branch 'feature/GH-8704-nextgen-units' of https://github.com/SAP/cloud-commerce-spartacus-storefront into feature/GH-8704-nextgen-units

Matejk00 committed:
Merge branch 'develop' into feature/GH-8704-nextgen-units

Matejk00 committed:
Merge branch 'develop' into feature/GH-8704-nextgen-units

Pucek9 committed:
Merge branch 'develop' into feature/GH-8704-nextgen-units

Pucek9 committed:
fix selector

Pucek9 committed:
up styles

Pucek9 committed:
up

Pucek9 committed:
up

Pucek9 committed:
Merge branch 'develop' of https://github.com/SAP/cloud-commerce-spartacus-storefront into feature/GH-8704-nextgen-units

Pucek9 committed:
Merge branch 'develop' of https://github.com/SAP/cloud-commerce-spartacus-storefront into feature/GH-8704-nextgen-units

Pucek9 committed:
fix unit

Pucek9 committed:
fix units

Pucek9 committed:
Merge branch 'develop' of https://github.com/SAP/cloud-commerce-spartacus-storefront into feature/GH-8704-nextgen-units

Pucek9 committed:
Merge branch 'develop' of https://github.com/SAP/cloud-commerce-spartacus-storefront into feature/GH-8704-nextgen-units

# Conflicts:
#	feature-libs/my-account/organization/styles/_list.scss

tobi-or-not-tobi committed:
fix router errors during ghost design

tobi-or-not-tobi committed:
fix tree toggling

tobi-or-not-tobi committed:
cx-views > 0 should only take the necessary vertical space

Pucek9 committed:
Merge branch 'develop' into feature/GH-8704-nextgen-units
